### PR TITLE
VC-4216 processing account migrate

### PR DESCRIPTION
**Pull request by alex-bodnar**
## Description

Closes: #XXXX

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: modules/vipcoin/chain/accounts/msg_server_account_migrate.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+			if xtypes.GetSDKAddress(a.Address).Equals(xtypes.GetSDKAddress(oldAddr)) {
+				// Update address to new one
+				a.Address = msg.Address
+				affiliateAcc.Affiliates[i] = a
+				if err := m.accountRepo.UpdateAccounts(affiliateAcc); err != nil {
+					return err
+				}
+				// Break loop
+				break
+			}
```
Comment (DirusK) [inline modules/vipcoin/chain/accounts/msg_server_account_migrate.go]:
```suggestion
		for idx := range affiliateAcc.Affiliates {
			if xtypes.GetSDKAddress(a.Address).Equals(xtypes.GetSDKAddress(oldAddr)) {
				// Update address to new one
				affiliateAcc.Affiliates[i].Address = msg.Address
				if err := m.accountRepo.UpdateAccounts(affiliateAcc); err != nil {
					return err
				}
				// Break loop
				break
			}
```

File: modules/vipcoin/chain/accounts/msg_server_account_migrate.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	juno "github.com/forbole/juno/v2/types"
+)
+
+// handleMsgAccountMigrate allows to properly handle a handleMsgAccountMigrate
+func (m *Module) handleMsgAccountMigrate(tx *juno.Tx, index int, msg *types.MsgAccountMigrate) error {
+	if err := m.accountRepo.SaveAccountMigrate(msg); err != nil {
+		return err
+	}
+
+	accArr, err := m.accountRepo.GetAccounts(filter.NewFilter().SetArgument(FieldHash, msg.Hash))
```
Comment (DirusK) [inline modules/vipcoin/chain/accounts/msg_server_account_migrate.go]: не экономь буквы  в переменных
лучше написать больше, но понятнее

File: modules/vipcoin/chain/accounts/msg_server_account_migrate.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	oldAddr := acc.Address
+	// change address and public key in account
+	acc.Address = msg.Address
+	acc.PublicKey = pkAny
+
+	// Change address in affiliates
+	for _, affiliate := range acc.Affiliates {
+		// Get affiliate account
+		accFilter := filter.NewFilter().SetArgument(FieldAddress, xtypes.GetSDKAddress(affiliate.Address))
+		arrAffiliateAcc, err := m.accountRepo.GetAccounts(accFilter)
```
Comment (DirusK) [inline modules/vipcoin/chain/accounts/msg_server_account_migrate.go]:
```suggestion
		arrAffiliateAcc, err := m.accountRepo.GetAccounts(filter.NewFilter().SetArgument(FieldAddress, xtypes.GetSDKAddress(affiliate.Address)))
```

File: _tests/database/accounts/account_migrate_test.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+	tests := []struct {
+		name    string
+		args    args
+		want    []*accountstypes.MsgAccountMigrate
+		wantErr bool
+	}{
+		{
+			name: "valid",
+			args: args{
+				accfilter: filter.NewFilter().SetArgument("address", "vcg1hwxejcutmafuedd8trjqumfdkst2498pggx45q"),
```
Comment (DirusK) [inline _tests/database/accounts/account_migrate_test.go]: используй константу для поля адрес

File: database/vipcoin/chain/accounts/account_migrate.go [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+	query := `INSERT INTO vipcoin_chain_accounts_account_migrate 
+			(creator, address, hash, public_key) 
+		VALUES 
+			(:creator, :address, :hash, :public_key)`
+
+	for _, account := range msg {
+		if _, err := tx.NamedExec(query, toAccountMigrateDatabase(account)); err != nil {
+			return err
+		}
+	}
```
Comment (DirusK) [inline database/vipcoin/chain/accounts/account_migrate.go]: Можно использовать PrepareNamedContext который подготовит кверю на следующие запросы  
Возможно это будет эффективнее за счет того что квери не повторяется

File: modules/vipcoin/chain/accounts/msg_server_account_migrate.go [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+	pkAny, err := types.PubKeyToAny(publicKey)
+	if err != nil {
+		return types.ErrInvalidPublicKeyField
+	}
+
+	oldAddr := acc.Address
+	// change address and public key in account
+	acc.Address = msg.Address
+	acc.PublicKey = pkAny
```
Comment (DirusK) [inline modules/vipcoin/chain/accounts/msg_server_account_migrate.go]:
```suggestion
	acc.PublicKey, err := types.PubKeyToAny(publicKey)
	if err != nil {
		return types.ErrInvalidPublicKeyField
	}

	oldAddr := acc.Address
	// change address and public key in account
	acc.Address = msg.Address
```

**Comment by IamStubborN** [comment]
миграция аккаунта (для текущего состояния из accounts):
- берем по hash существующий аккаунт
- изменяем переданные атрибуты из месседжа
- сохраняем аккаунт

**Comment by IamStubborN** [comment]
squash commits

File: database/vipcoin/chain/accounts/accounts.go [hunk 7]
```diff
@@ -109,14 +111,16 @@ func (r Repository) UpdateAccounts(accounts ...*accountstypes.Account) error {
...
 
 	query := `UPDATE vipcoin_chain_accounts_accounts SET
-				 hash = :hash, public_key = :public_key, kinds = :kinds,
+				 address = :address, public_key = :public_key, kinds = :kinds,
 				 state = :state, extras = :extras, affiliates = :affiliates, wallets = :wallets
-			 WHERE address = :address`
+			 WHERE hash = :hash`
 
-	queryAffiliates := "SELECT affiliates FROM vipcoin_chain_accounts_accounts WHERE (address IN ($1))"
+	queryAffiliates := "SELECT affiliates FROM vipcoin_chain_accounts_accounts WHERE (hash IN ($1))"
```
Comment (IamStubborN) [inline database/vipcoin/chain/accounts/accounts.go]: WHERE hash = $1?

**Review by IamStubborN — APPROVED** [review]
LGTM

File: modules/vipcoin/chain/accounts/msg_server_account_migrate.go [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+	if len(acc) != 1 {
+		return types.ErrInvalidHashField
+	}
+
+	publicKey, err := types.PubKeyFromString(msg.PublicKey)
+	if err != nil {
+		return types.ErrInvalidPublicKeyField
+	}
+
+	pkAny, err := types.PubKeyToAny(publicKey)
```
Comment (DirusK) [inline modules/vipcoin/chain/accounts/msg_server_account_migrate.go]:
```suggestion
	acc[0].PublicKey , err = types.PubKeyToAny(publicKey)
```

**Comment by DirusK** [comment]
LGTM

**Pull request merged**